src/man/pkgdep.1.txt
User Commands                                             pkgdep(1)


NAME
     pkgdep - image packaging system dependency analyzer

SYNOPSIS
     /usr/bin/pkgdep [options] command [cmd_options] [operands]

     /usr/bin/pkgdep generate [-IMm] manifest_path proto_dir

     /usr/bin/pkgdep resolve [-dMos] manifest_path ...

DESCRIPTION
     The pkgdep command allows the user to produce the files upon which a
     package depends due to its contents, or resolve those dependencies into
     the packages which contain the files depended upon. The typical use is to
     use the production phase to independently generate the dependencies for
     all packages being published, then pass all those manifests to the
     resolve phase so it can resolve any dependencies between the packages
     being published as well as resolving against the reference set of 
     packages. Currently, the reference set of packages is defined as the
     packages currently installed on the publisher's system.

OPTIONS
     The following options are supported:

     -R dir
          Operate on the image rooted at dir, rather than the one discovered
          automatically.

     --help or -?
          Displays a usage message.

SUBCOMMANDS
     The following subcommands are supported:

     generate [-IMm] manifest_path proto_dir
          Produce the dependencies on files of the manifest specified by 
          manifest_path, whose contents are found in proto_dir.

          With -I, the dependencies which are satisified within the manifest
          given will be shown.

          With -M, a list of those file types which could not be analyzed 
          will be displayed. 

          With -m, the original manifest will be repeated with any 
          discovered dependencies added after.

     resolve [-mo] [-d output_dir] [-s suffix] manifest_path ...
          Transform dependencies on files into dependencies on the packages
          which deliver those files.  Dependencies are first resolved against
          the manifests given on the command line and then against the packages
          installed on the system.  By default, the dependencies for each 
          manifest are placed in a file whose name is determined by appending
          ".res" to the manifest's path.

          With -m, repeat the manifest, with any dependencies produced by
          the generate step removed, before adding the resolved 
          dependencies.

          With -o, write the results to standard out.

          With -d, write the resolved dependencies for each manifest 
          provided in a separate file in output_dir.  By default, each file
          will have the same basename as the manifest that was the source
          of the dependencies written to that file.

          With -s, for each output file, append the given suffix (a "." 
          will be added to the argument if not provided) to the basename of 
          the file that was the source of the resolved dependencies.


EXAMPLES
     Example 1: Generate the dependencies for the manifest written in foo,
     whose content directory is in ./bar/baz and store the results in 
     foo.fdeps.

     $ pkgdep generate foo ./bar/baz > foo.fdeps

     Example 2: Resolve the file dependencies in foo.fdeps, and bar.fdeps 
     against each other, and the packages currently installed on the system.

     $ pkgdep resolve foo.fdeps bar.fdeps
     $ ls *.res
       foo.fdeps.res     bar.fdeps.res

     Example 3: Generate the file dependencies for two manifests (foo and 
     bar) and retain all the information in the original manifests. Then 
     resolve the file dependencies and place the resulting manifests, which
     could be used with pkgsend publish, in ./res

     $ pkgdep generate -m foo > ./deps/foo
     $ pkgdep generate -m bar > ./deps/bar
     $ pkgdep resolve -m -d ./res ./deps/foo ./deps/bar
     $ ls ./res
       foo     bar


doc/history.txt

pkg
HISTORY

1. Summary

The intent of history is to enable pkg clients to maintain a record of all
image-modifying operations that they perform.

2. Discussion

2.1 Operations

    The following operations are currently recorded by the clients that are
    part of pkg(5):

        add-publisher
        image-create
        image-set-attributes
        image-update
        install
        purge-history
        rebuild-index
        refresh-publisher
        remove-publisher
        set-publisher
        set-preferred-publisher
        uninstall
        update-publisher

    These operations were chosen because they have the potential to alter the
    behavior of pkg(5) operations or because they modify an image.

3. History Entries

3.1 Information Recorded

    Each history entry contains a record of basic information about the client
    performing the operation, along with a mix of required and optional
    information about the operation performed.

    The following information about the client is always recorded:

        client_name     - The name of the client performing the operation
                          (e.g. pkg, packagemanager)

        client_version  - The value of pkg.VERSION at the time the client
                          performed the operation (e.g. 2e5300c4f0a4+)

        client_args     - The command-line arguments used when executing the
                          client (e.g. /usr/bin/pkg install foo)

    The following information about the operation performed is always recorded:

        operation_name  - The name of the operation performed (refer to 2.1)

        start_time      - When the operation started (e.g. 20080916T214726Z)

        end_time        - When the operation ended (e.g. 20080916T214800Z)

        userid          - The id of the user that performed the operation
                          (e.g. 0)

        username        - The username of the user that performed the operation
                          (e.g. root)

        result          - The outcome of the operation and the reason for it
                          (e.g. Failed, Transport)

    The following information about the operation will be recorded if provided
    by the client or api:

        start_state     - Information about the operation requested before any
                          evaluation of that request is performed (e.g. an
                          image plan before evaluation)

        end_state       - Information about the operation requested after
                          evaluation of that request has been performed (e.g.
                          image plan after evaluation)

        errors          - Any errors that were encountered while performing the
                          operation (i.e. tracebacks, exceptions, etc.)

3.2 Storage

    Each history entry is recorded in a XML file located in the metadata
    directory of an image (e.g. /var/pkg) in a directory named "history".
    Each XML file is named after the pattern %Y%m%dT%H%M%SZ-sequence.xml.
    Where sequence is a numeric value appended so that when multiple
    operations are performed on an image within the same second (e.g. -02,
    -03, etc.) entries are still written correctly.

3.3 File Format

    It should be noted that this format description is that of a private
    interface and is subject to change.

    History entries are recorded in a XML file with a fairly simplistic
    structure.  The format is as follows:

        The first line of each file is a standard xml header along with the
        encoding used to save the information.  This allows the pkg history
        command to correctly display this information if the client changes
        locales later.

        <?xml version="1.0" encoding="ascii"?>

        The second element of every history XML file is a root element used to
        contain the client and operation information.  Only one per file ever
        occurs.

        <history>

            The client element has a name and version attribute used to record
            client_name and client_version.  Only one per file ever occurs.

            <client name="pkg" version="2e5300c4f0a4+">

                The args element contains one or more "arg" (argument) elements
                that match each element of the system argument list used to
                execute the client.  It has no attributes.  Only one per file
                ever occurs.

                <args>

                    Each arg element contains a CDATA element to encapsulate
                    the raw information for the argument.  It has no attributes.

                    <arg>
                        <![CDATA[/usr/bin/pkg]]>
                    </arg>
                    <arg>
                        <![CDATA[-R]]>
                    </arg>
                    <arg>
                        <![CDATA[/tmp/test-image]]>
                    </arg>
                    <arg>
                        <![CDATA[install]]>
                    </arg>
                    <arg>
                        <![CDATA[SUNWvim]]>
                    </arg>
                </args>
            </client>

            The operation element has the following attributes: name,
            start_time, end_time, userid, and username.  It can also
            contain the following, optional elements: start_state,
            end_state, and errors.  It only occurs once per file.

            <operation end_time="20080912T225513Z" name="install" result="Succeeded"
                start_time="20080912T225327Z" userid="101" username="swalker">

                The start_state element is used to store information about the
                operation performed before a request is evaluated (e.g. image
                plan before evaluation).  It always contains a CDATA element with
                the related information and only occurs once per file.  It has no
                attributes.

                <start_state>

src/pkgdep.py
#!/usr/bin/python2.4
#
# CDDL HEADER START
#
# The contents of this file are subject to the terms of the
# Common Development and Distribution License (the "License").
# You may not use this file except in compliance with the License.
#
# You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
# or http://www.opensolaris.org/os/licensing.
# See the License for the specific language governing permissions
# and limitations under the License.
#
# When distributing Covered Code, include this CDDL HEADER in each
# file and include the License file at usr/src/OPENSOLARIS.LICENSE.
# If applicable, add the following below this CDDL HEADER, with the
# fields enclosed by brackets "[]" replaced with your own identifying
# information: Portions Copyright [yyyy] [name of copyright owner]
#
# CDDL HEADER END
#

#
# Copyright 2009 Sun Microsystems, Inc.  All rights reserved.
# Use is subject to license terms.
#

import errno
import getopt
import gettext
import locale
import os
import sys
import traceback

import pkg
import pkg.actions as actions
import pkg.client.api as api
import pkg.client.api_errors as api_errors
import pkg.client.progress as progress
import pkg.misc as misc
import pkg.publish.dependencies as dependencies
from pkg.misc import msg, emsg, PipeError

CLIENT_API_VERSION = 19
PKG_CLIENT_NAME = "pkgdep"

DEFAULT_SUFFIX = ".res"

def error(text, cmd=None):
        """Emit an error message prefixed by the command name """

        if cmd:
                text = "%s: %s" % (cmd, text)
        else:
                # If we get passed something like an Exception, we can convert
                # it down to a string.
                text = str(text)

        # If the message starts with whitespace, assume that it should come
        # *before* the command-name prefix.
        text_nows = text.lstrip()
        ws = text[:len(text) - len(text_nows)]

        # This has to be a constant value as we can't reliably get our actual
        # program name on all platforms.
        emsg(ws + "pkgdep: " + text_nows)

def usage(usage_error=None, cmd=None, retcode=2):
        """Emit a usage message and optionally prefix it with a more specific
        error message.  Causes program to exit."""

        if usage_error:
                error(usage_error, cmd=cmd)
        emsg (_("""\
Usage:
        pkgdep [options] command [cmd_options] [operands]

Subcommands:
        pkgdep generate [-IMm] manifest proto_dir
        pkgdep [options] resolve [-dMos] manifest ...

Options:
        -R dir
        --help or -?
Environment:
        PKG_IMAGE"""))

        sys.exit(retcode)

def generate(args):
        """Produce a list of file dependencies from a manfiest and a proto
        area."""
        try:
                opts, pargs = getopt.getopt(args, "IMm?",
                    ["help"])
        except getopt.GetoptError, e:
                usage(_("illegal global option -- %s") % e.opt)

        remove_internal_deps = True
        echo_manf = False
        show_missing = False
        show_usage = False

        for opt, arg in opts:
                if opt == "-I":
                        remove_internal_deps = False
                elif opt == "-m":
                        echo_manf = True
                elif opt == "-M":
                        show_missing = True
                elif opt in ("--help", "-?"):
                        show_usage = True
        if show_usage:
                usage(retcode=0)
        if len(pargs) != 2:
                usage()

        retcode = 0
                
        manf = pargs[0]
        proto_dir = pargs[1]

        try:
                ds, es, ms = dependencies.list_implicit_deps(manf, proto_dir,
                    remove_internal_deps)
        except IOError, e:
                if e.errno == errno.ENOENT:
                        error("Could not find manifest file %s" % manf)
                        return 1
                raise

        if echo_manf:
                fh = open(manf, "rb")
                for l in fh:
                        msg(l.rstrip())
                fh.close()
        
        for d in sorted(ds):
                msg(d)

        if show_missing:
                for m in ms:
                        emsg(m)
                
        for e in es:
                emsg(e)
                retcode = 1
        return retcode

def resolve(args, img_dir):
        """Take a list of manifests and resolve any file dependencies, first
        against the other published manifests and then against what is installed
        on the machine."""
        out_dir = None
        echo_manifest = False
        output_to_screen = False
        suffix = None
        opts, pargs = getopt.getopt(args, "d:mos:")
        for opt, arg in opts:
                if opt == "-d":
                        out_dir = arg
                elif opt == "-m":
                        echo_manifest = True
                elif opt == "-o":
                        output_to_screen = True
                elif opt == "-s":
                        suffix = arg

        if (out_dir or suffix) and output_to_screen:
                usage(_("-o cannot be used with -d or -s"))

        manifest_paths = [os.path.abspath(fp) for fp in pargs]

        if out_dir:
                out_dir = os.path.abspath(out_dir)

        if img_dir is None:
                try:
                        img_dir = os.environ["PKG_IMAGE"]
                except KeyError:
                        try:
                                img_dir = os.getcwd()
                        except OSError, e:
                                try:
                                        img_dir = os.environ["PWD"]
                                        if not img_dir or img_dir[0] != "/":
                                                img_dir = None
                                except KeyError:
                                        img_dir = None

        if img_dir is None:
                error(_("Could not find image.  Use the -R option or set "
                    "$PKG_IMAGE to point\nto an image, or change the working "
                    "directory to one inside the image."))
                return 1

        # Becuase building an ImageInterface permanently changes the cwd for
        # python, it's necessary to do this step after resolving the paths to
        # the manifests.
        try:
                api_inst = api.ImageInterface(img_dir, CLIENT_API_VERSION,
                    progress.QuietProgressTracker(), None, PKG_CLIENT_NAME)
        except api_errors.ImageNotFoundException, e:
                error(_("'%s' is not an install image") % e.user_dir)
                return 1                
        pkg_deps, errs = dependencies.resolve_deps(manifest_paths, api_inst)
        ret_code = 0
        
        if output_to_screen:
                ret_code = pkgdeps_to_screen(pkg_deps, manifest_paths,
                    echo_manifest)
        elif out_dir:
                ret_code = pkgdeps_to_dir(pkg_deps, manifest_paths, out_dir,
                    suffix, echo_manifest)
        else:
                ret_code = pkgdeps_in_place(pkg_deps, manifest_paths, suffix,
                    echo_manifest)

        for path, file_dep, pvars in errs:
                if ret_code == 0:
                        ret_code = 1
                emsg("%s has unresolved dependency '%s' under the following "
                     "combinations of variants:" % (path, file_dep))
                for grp in pvars.get_unsatisfied():
                        emsg(" ".join([
                            ("%s:%s" % (name, val)) for name, val in grp]))
        return ret_code

def echo_line(l):
        """Given a line from a manifest, determines whether that line should
        be repeated in the output file if echo manifest has been set."""

        try:
                act = actions.fromstr(l.rstrip())
        except KeyboardInterrupt:
                raise
        except actions.ActionError:
                return True
        else:
                return not dependencies.is_file_dependency(act)

def explode(dep_with_variantsets):
        sat_tups = dep_with_variantsets.get_variants().get_satisfied()
        if sat_tups is None:
                return dep_with_variantsets
        res = []
        for tup in sat_tups:
                attrs = dep_with_variantsets.attrs.copy()
                attrs.update(dict(tup))
                res.append(str(actions.depend.DependencyAction(**attrs)))
        return "\n".join(res).rstrip()
        
def pkgdeps_to_screen(pkg_deps, manifest_paths, echo_manifest):
        """Write the resolved package dependencies to stdout.

        'pkg_deps' is a dictionary that maps a path to a manifest to the
        dependencies that were resolved for that manifest.

        'manifest_paths' is a list of the paths to the manifests for which
        file dependencies were resolved.

        'echo_manifest' is a boolean which determines whether the original
        manifest will be written out or not."""

        ret_code = 0
        for p in manifest_paths:
                msg(p)
                if echo_manifest:
                        try:
                                fh = open(p, "rb")
                                for l in fh:
                                        if echo_line(l):
                                                msg(l.rstrip())
                                fh.close()
                        except EnvironmentError:
                                emsg(_("Could not open %s to echo manifest") %
                                    p)
                                ret_code = 1
                for d in pkg_deps[p]:
                        msg(explode(d))
                msg(_("\n\n"))
        return ret_code

def write_res(deps, out_file, echo_manifest, manifest_path):
        """Write the dependencies resolved, and possibly the manifest, to the
        destination file.

        'deps' is a list of the resolved dependencies.

        'out_file' is the path to the destination file.

        'echo_manifest' determines whether to repeat the original manifest in
        the destination file.

        'manifest_path' the path to the manifest which generated the
        dependencies."""

        ret_code = 0
        try:
                out_fh = open(out_file, "wb")
        except EnvironmentError:
                ret_code = 1
                emsg(_("Could not open output file %s for writing") %
                    out_file)
                return ret_code
        if echo_manifest:
                try:
                        fh = open(manifest_path, "rb")
                except EnvironmentError:
                        ret_code = 1
                        emsg(_("Could not open %s to echo manifest") %
                            manifest_path)
                for l in fh:
                        if echo_line(l):
                                out_fh.write(l)
                fh.close()
        for d in deps:
                out_fh.write("%s\n" % explode(d))
        out_fh.close()
        return ret_code

def pkgdeps_to_dir(pkg_deps, manifest_paths, out_dir, suffix, echo_manifest):
        """Given an output directory, for each manifest given, writes the
        dependencies resolved to a file in the output directory.

        'pkg_deps' is a dictionary that maps a path to a manifest to the
        dependencies that were resolved for that manifest.

        'manifest_paths' is a list of the paths to the manifests for which
        file dependencies were resolved.

        'out_dir' is the path to the directory into which the dependency files
        should be written.

        'suffix' is the string to append to the end of each output file.

        'echo_manifest' is a boolean which determines whether the original
        manifest will be written out or not."""

        ret_code = 0
        if not os.path.exists(out_dir):
                try:
                        os.makedirs(out_dir)
                except EnvironmentError, e:
                        emsg(_("Out dir %s does not exist and could not be "
                            "created. Error is: %s") % e)
                        return 1
        if suffix and suffix[0] != ".":
                suffix = "." + suffix
        for p in manifest_paths:
                out_file = os.path.join(out_dir, os.path.basename(p))
                if suffix:
                        out_file += suffix
                tmp_rc = write_res(pkg_deps[p], out_file, echo_manifest, p)
                if not ret_code:
                        ret_code = tmp_rc
        return ret_code

def pkgdeps_in_place(pkg_deps, manifest_paths, suffix, echo_manifest):
        """Given an output directory, for each manifest given, writes the
        dependencies resolved to a file in the output directory.

        'pkg_deps' is a dictionary that maps a path to a manifest to the
        dependencies that were resolved for that manifest.

        'manifest_paths' is a list of the paths to the manifests for which
        file dependencies were resolved.

        'out_dir' is the path to the directory into which the dependency files
        should be written.

        'suffix' is the string to append to the end of each output file.

        'echo_manifest' is a boolean which determines whether the original
        manifest will be written out or not."""

        ret_code = 0
        if not suffix:
                suffix = DEFAULT_SUFFIX
        if suffix[0] != ".":
                suffix = "." + suffix
        for p in manifest_paths:
                out_file = p + suffix
                tmp_rc = write_res(pkg_deps[p], out_file, echo_manifest, p)
                if not ret_code:
                        ret_code = tmp_rc
        return ret_code

def main_func():
        misc.setlocale(locale.LC_ALL, "", error)
        gettext.install("pkg", "/usr/share/locale")

        try:
                opts, pargs = getopt.getopt(sys.argv[1:], "R:?",
                    ["help"])
        except getopt.GetoptError, e:
                usage(_("illegal global option -- %s") % e.opt)

        show_usage = False
        img_dir = None
        for opt, arg in opts:
                if opt == "-R":
                        img_dir = arg
                elif opt in ("--help", "-?"):
                        show_usage = True

        subcommand = None
        if pargs:
                subcommand = pargs.pop(0)
                if subcommand == "help":
                        show_usage = True

        if show_usage:
                usage(retcode=0)
        elif not subcommand:
                usage()

        if subcommand == "generate":
                if img_dir:
                        usage(_("generate subcommand doesn't use -R"))
                return generate(pargs)
        elif subcommand == "resolve":
                return resolve(pargs, img_dir)
        else:
                usage(_("unknown subcommand '%s'") % subcommand)

#
# Establish a specific exit status which means: "python barfed an exception"
# so that we can more easily detect these in testing of the CLI commands.
#
if __name__ == "__main__":
        try:
                __ret = main_func()
        except api_errors.MissingFileArgumentException, e:
                error("The manifest file %s could not be found." % e.path)
                __ret = 1
        except RuntimeError, _e:
                emsg("%s: %s" % (PKG_CLIENT_NAME, _e))
                __ret = 1
        except (PipeError, KeyboardInterrupt):
                # We don't want to display any messages here to prevent
                # possible further broken pipe (EPIPE) errors.
                __ret = 1
        except SystemExit, _e:
                raise _e
        except:
                traceback.print_exc()
                error(
                    _("\n\nThis is an internal error.  Please let the "
                    "developers know about this\nproblem by filing a bug at "
                    "http://defect.opensolaris.org and including the\nabove "
                    "traceback and this message.  The version of pkg(5) is "
                    "'%s'.") % pkg.VERSION)
                __ret = 99
        sys.exit(__ret)
